[pjlinkdevice] PJLink Binding initial contribution

nils opened the pull request:
This is the initial version of a [PJLink](https://pjlink.jbmia.or.jp/english/) device binding. It allows to control devices which implement the standard, in particular digital projectors. Please see the README.md file for some basic information about the binding.

~~Currently it has some limitations, in particular only the power on/off control is implemented, as this is sufficient for my own usecase~~

~~If there is a chance that this binding is merged into the main repository, I'd consider adding the additional (PJLink Class 1) features like authentication, volume setting, input source selection.~~
→ See update below.

nils commented:
Hi,

I've completed PJLink class 1 support including authentication.

I'd be happy to receive some feedback about the binding. Did I forget something that avoids that?

Most importantly, I'm interested whether there is a chance that this pull request is merged (after incorporating potential feedback, of course)?

Nils

nils commented:
I just added polling to be able to react on externally triggered state changes (e.g. user using the projector remote to switch on/off, change input channel etc.)

Also, I did another rebase to current master.

wborn commented:
> I'd be happy to receive some feedback about the binding. Did I forget something that avoids that?

Thanks for the binding! There's nothing that prevents this from being reviewed. What helps is to only add new commits instead of force pushing a single squashed commit. That has less impact on pending reviews. It will also be more clear to reviewers what has changed. We can squash the commits while merging the PR.

Hilbrand commented on addons/binding/org.openhab.binding.pjlinkdevice/README.md:
```diff
@@ -0,0 +1,50 @@
+
+The *pjLinkDevice* thing type has the following parameters:
+
+- *ipAddress*: the IPv4 address of the device
```
Use table notation for configuration and channels. It's more readable:
```
| Parameter | Description                    |
|-----------|--------------------------------|
| ipAddress | the IPv4 address of the device |
```
Make sure to the bars are all aligned vertically in the raw text.

nils replied:
done

Hilbrand commented on addons/binding/org.openhab.binding.pjlinkdevice/about.html:
```diff
@@ -0,0 +1,33 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
```
Replace with NOTICE file : https://community.openhab.org/t/tutorial-update-license-to-epl-v2-0-of-your-binding-in-development-or-pull-request/65088

nils replied:
done

nils commented:
Thanks @wborn for some initial feedback. I'll add commits from now on instead of amending.

Hilbrand commented on addons/binding/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,302 @@
+@NonNullByDefault
+public class PJLinkDeviceHandler extends BaseThingHandler {
+
+    public class RefreshRunnable implements Runnable {
```
Prefer to write these as normal methods or as field. In the scheduler you can simply use `this::refresh` Where `refresh` is the name of the method or as field:
```
private final Runnable refreshRunnable = () -> {
  <statements>
};
```

nils replied:
Nice! And done.

Hilbrand commented on addons/binding/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/device/PJLinkDevice.java:
```diff
@@ -0,0 +1,321 @@
+        this(tcpPort, ipAddress, adminPassword, TIMEOUT);
+    }
+
+    protected Logger getLogger() {
```
Why not have a field:`private final Logger logger = LoggerFactory.getLogger(PJLinkDevice.class);`

nils replied:
I think that was a remained from when I included the device address in the logger scope. Now it was just useless and is replaced by the field.

Hilbrand commented on addons/binding/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/device/command/ErrorCode.java:
```diff
@@ -0,0 +1,52 @@
+
+    public String getText() throws ResponseException {
+        final HashMap<ErrorCode, String> texts = new HashMap<ErrorCode, String>();
+        texts.put(UNDEFINED_COMMAND, "Undefined command");
```
This should be a parameter to the enum:
`UNDEFINED_COMMAND("Undefined command")` and then `getText()` simply returns the value set in a field:.

nils replied:
Nice! That did not feel quite right for me too...
Done

Hilbrand commented on addons/binding/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/discovery/AbstractDiscoveryParticipant.java:
```diff
@@ -0,0 +1,119 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableSet;
```
please don't use `com.google.common.collect` this is part of the large guava library and we're trying to get rid of it as it's only used a very small set.

nils replied:
done

nils commented:
Thanks a lot for the review! I've tried my best to follow your advice, looking forward to the next round! :+1: 
Are you going to press "resolve conversation" or should I do that?

lolodomo commented:
I will make a full review.
As first remark, I have the feeling you made only a partial move to the new build system. As example, the MANIFEST.MF file should have been suppressed.

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,82 @@
+			</parameter>
+			<parameter name="refreshPower" type="boolean" required="true">
+				<default>false</default>
+				<label>Enable polling for power state</label>
```
Maybe define a short label + a description.

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,82 @@
+			</parameter>
+			<parameter name="refreshMute" type="boolean" required="true">
+				<default>false</default>
+				<label>Enable polling for mute state</label>
```
Maybe define a short label + a description.

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,82 @@
+			</parameter>
+			<parameter name="refreshInputChannel" type="boolean" required="true">
+				<default>false</default>
+				<label>Enable polling for selected input channel</label>
```
Maybe define a short label + a description.

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/README.md:
```diff
@@ -0,0 +1,58 @@
+|-----------------------|----------------------------------------------------------------------------------------------------------------------|
+| ipAddress             | the IPv4 address of the device                                                                                       |
+| tcpPort               | the listening TCP port of the devices                                                                                |
+| password              | the PJLink password of the device (should be left empty for devices without authentication)                          |
```
Parameter is in fact "adminPassword".

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,82 @@
+
+	</thing-type>
+
+	<channel-type id="powerChannelType">
```
"power" would be sufficient.

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,82 @@
+	</channel-type>
+
+
+	<channel-type id="inputChannelType">
```
"input" would be sufficient.

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,82 @@
+		<description>Select the input signal to use</description>
+	</channel-type>
+
+	<channel-type id="audioMuteChannelType">
```
"audioMute" would be sufficient.

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,82 @@
+		<description>Select the audio mute status</description>
+	</channel-type>
+
+	<channel-type id="videoMuteChannelType">
```
"videoMute" would be sufficient.

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,82 @@
+		<label>Power</label>
+		<description>Power ON/OFF the projector</description>
+	</channel-type>
+
```
Please suppress one blank line.

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,82 @@
+		<description>A PJLink compatible device, e.g. a digital projector</description>
+
+		<channels>
+			<channel id="powerChannel" typeId="powerChannelType" />
```
Id "power" and type "power" would be sufficient.

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,82 @@
+
+		<channels>
+			<channel id="powerChannel" typeId="powerChannelType" />
+			<channel id="inputChannel" typeId="inputChannelType" />
```
Id "input" and type "input" would be sufficient.

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,82 @@
+		<channels>
+			<channel id="powerChannel" typeId="powerChannelType" />
+			<channel id="inputChannel" typeId="inputChannelType" />
+			<channel id="audioMuteChannel" typeId="audioMuteChannelType" />
```
Id "audioMute" and type "audioMute" would be sufficient.

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,82 @@
+			<channel id="powerChannel" typeId="powerChannelType" />
+			<channel id="inputChannel" typeId="inputChannelType" />
+			<channel id="audioMuteChannel" typeId="audioMuteChannelType" />
+			<channel id="videoMuteChannel" typeId="videoMuteChannelType" />
```
Id "videoMute" and type "videoMute" would be sufficient.

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/pom.xml:
```diff
@@ -0,0 +1,16 @@
+  <artifactId>org.openhab.binding.pjlinkdevice</artifactId>
+
+  <name>openHAB Add-ons :: Bundles :: PJLink Binding</name>
+
```
Your dependency is very probably missing here.

nils replied:
Is this the right way of doing it?

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,82 @@
+				<label>Password</label>
+				<description>The password of the PJLink device.</description>
+			</parameter>
+			<parameter name="tcpPort" type="integer" required="true">
```
As there is a default value and as we can see in your example, this parameter is not required.

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,82 @@
+				<label>TCP Port</label>
+				<description>The TCP port of the PJLink device to control.</description>
+			</parameter>
+			<parameter name="refresh" type="integer" required="true" min="0">
```
As there is a default value and as we can see in your example, this parameter is not required.

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,82 @@
+				<label>Refresh interval (s)</label>
+				<description>How often to poll the device state (in seconds). A value of zero will disable polling.</description>
+			</parameter>
+			<parameter name="refreshPower" type="boolean" required="true">
```
As there is a default value and as we can see in your example, this parameter is not required.

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,82 @@
+				<default>false</default>
+				<label>Enable polling for power state</label>
+			</parameter>
+			<parameter name="refreshMute" type="boolean" required="true">
```
As there is a default value and as we can see in your example, this parameter is not required.

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,82 @@
+				<default>false</default>
+				<label>Enable polling for mute state</label>
+			</parameter>
+			<parameter name="refreshInputChannel" type="boolean" required="true">
```
As there is a default value and as we can see in your example, this parameter is not required.

nils commented:
Thanks @lolodomo for the review. I'll have a detailed look and work on it in the next days.

Concerning naming, I'm starting to get some doubts concerning the naming of the binding itself ("pjlinkdevice"). Would "pjlink" be a better name?

lolodomo commented:
I will finish my review soon. 

IMHO,  PjLink would be a better name,  yes,  clearly.

lolodomo commented:
Thinking about it, pjlinkprojector would be a better name, giving more meaning about what it is.

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,313 @@
+            updateProperty(PJLinkDeviceBindingConstants.PARAMETER_MANUFACTURER, e.toString());
+        }
+        try {
+            updateProperty(PJLinkDeviceBindingConstants.PARAMETER_MODEL, device.getModel());
```
You can rather use the existing system property, Thing.PROPERTY_MODEL_ID

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,313 @@
+            updateProperty(PJLinkDeviceBindingConstants.PARAMETER_NAME, e.toString());
+        }
+        try {
+            updateProperty(PJLinkDeviceBindingConstants.PARAMETER_MANUFACTURER, device.getManufacturer());
```
You can rather use the existing system property, Thing.PROPERTY_VENDOR

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,313 @@
+            }
+        } catch (ResponseException e) {
+            // okay, cannot retrieve class information
+            updateProperty(PJLinkDeviceBindingConstants.PARAMETER_ERROR_STATUS, e.toString());
```
Looks weird to set the property with the exception string.
Why not just keeping the property unset in this case ?

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,313 @@
+            updateProperty(PJLinkDeviceBindingConstants.PARAMETER_MODEL, e.toString());
+        }
+        try {
+            updateProperty(PJLinkDeviceBindingConstants.PARAMETER_COMBINED_ID, device.getFullDescription());
```
Is there a real interest to provide two different properties, PARAMETER_NAME and PARAMETER_COMBINED_ID ?

nils replied:
I thought it was handy for diagnosis, but I suppose the users can also copy three fields (manufacturer, model and name). I have removed the property.

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,313 @@
+            for (Map.Entry<ErrorStatusDevicePart, ErrorStatusQueryResponseState> entry : errorStatus.entrySet()) {
+                String key = entry.getKey().getCamelCaseText();
+                String value = entry.getValue().getText();
+                updateProperty(PJLinkDeviceBindingConstants.PARAMETER_ERROR_STATUS + key, value);
```
I don't see real interest to provide that as property for the end user. This is more technical information.

nils replied:
I am an end user and I'm interested in this technical information :-)

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,313 @@
+            updateProperty(PJLinkDeviceBindingConstants.PARAMETER_ERROR_STATUS, e.toString());
+        }
+        try {
+            updateProperty(PJLinkDeviceBindingConstants.PARAMETER_LAMP_HOURS, device.getLampHours());
```
This should not be a property but an information to put in a channel.

nils replied:
The issue I see with that is that a PJLink projector does not necessarily have only one lamp. 
The string then looks something like 
```
8259 1
13448 0
...
```
which means that lamp one has 8259 hours usage time and is active, lamp two has 13448 hours and is not active, ...
The only way I can imagine to have this as channels it to create channels dynamically based on the number of lamps present in the projector. Sounds like a rather overengineered solution for this (for most users) less crucial bit of functionality.

lolodomo replied:
I am not aware of such projectors with several lamps !
I imagine you could define 2 channels, one for lamp 1 and one for lamp 2.
One interest of having a channel is that the user could use this information in a rule.

lolodomo replied:
Better idea: just one channel with a channel parameter to set the lamp number. 
Maybe this feature is more to keep trace of duration for the replaced lamps?

nils replied:
I assumed the multi-lamp feature is intended for professional high-end projectors which might have redundant/backup lamps (for example in cinemas). In the standard specification it's not even limited to any number, it could even be 20 or more lamps (theoretically :-)). So the idea with the channel parameter sounds appropriate to me :+1: However that might take some time for me to implement (new concept for me)

You're absolutely right, for example users could use this information to know when a lamp comes close to its expected lifetime.

nils replied:
Actually there is a limit in the PJlink spec  :-)

> Maximum value of n is 8.

I've experimented with the channel parameter approach and it works in general, but some things like validation of the channel configuration (when and how to do it?) and how to property treat those additional configured channels in handleCommand are not clear to me.

I propose that I add another pull request after the binding is merged, what is your opinion on that?

nils replied:
Pull request added for this functionality: https://github.com/openhab/openhab2-addons/pull/6063

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,313 @@
+            updateProperty(PJLinkDeviceBindingConstants.PARAMETER_LAMP_HOURS, device.getLampHours());
+        } catch (ResponseException e) {
+            // okay, cannot retrieve class information
+            updateProperty(PJLinkDeviceBindingConstants.PARAMETER_LAMP_HOURS, e.toString());
```
Looks weird to set the property with the exception string.
Why not just keeping the property unset in this case ?

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,313 @@
+            updateProperty(PJLinkDeviceBindingConstants.PARAMETER_OTHER_INFORMATION, device.getOtherInformation());
+        } catch (ResponseException e) {
+            // okay, cannot retrieve class information
+            updateProperty(PJLinkDeviceBindingConstants.PARAMETER_OTHER_INFORMATION, e.toString());
```
Looks weird to set the property with the exception string.
Why not just keeping the property unset in this case ?

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,313 @@
+        }
+    }
+
+    private void updateDeviceProperties(PJLinkDevice device) throws IOException, AuthenticationException {
```
When you plan to update several properties, a more efficient pattern is:
```
        Map<String, String> properties = editProperties();
        properties.put(PROPERTY1, value1);
        properties.put(PROPERTY2, value2);
        updateProperties(properties);
```

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,313 @@
+            states.add(new StateOption(input.getPJLinkRepresentation(), input.getText()));
+        }
+
+        ChannelTypeUID channelTypeUID = new ChannelTypeUID(
```
You can totally replace all this by providing a DynamicStateDescriptionProvider.
And you don't need two different channels, inputChannel and inputChannelDynamic. inputChannel with dynamic options will be sufficient.

nils replied:
done

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,313 @@
+
+    private void handleAuthenticationException(AuthenticationException e) {
+        updateProperty(PJLinkDeviceBindingConstants.PARAMETER_AUTHENTICATION_REQUIRED, new Boolean(true).toString());
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.HANDLER_CONFIGURATION_PENDING, e.getMessage());
```
I think ThingStatusDetail.CONFIGURATION_ERROR should be used in this case with an additional message telling the user that he probably forgets to set a password.

nils replied:
changed to CONFIGURATION_ERROR;
e.getMessage() will be either "Authentication error, wrong password provided?" or "No password provided, but device requires authentication" in real life

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,313 @@
+    @Override
+    public void initialize() {
+        config = getConfigAs(PJLinkDeviceConfiguration.class);
+        setupDevice();
```
You should consider running the setupDevice in a thread as it will establish a connection with the projector and so it could take some time.

nils replied:
Could you point me to an example showing which flavour of thread creation is preferred in OpenHAB?

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,313 @@
+
+    public void refresh() {
+        // Do not poll if configuration is incomplete
+        if (PJLinkDeviceHandler.this.getThing().getStatusInfo()
```
Just test if thing is ONLINE

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,313 @@
+        logger.trace("Received command {} on channel {}", command, channelUID.getId());
+        try {
+            PJLinkDevice device = config.getDevice();
+            if (channelUID.getId().equals(CHANNEL_POWER)) {
```
The genreal used pattern is a switch on channelUID.getId()

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,313 @@
+                }
+            }
+            if (channelUID.getId().equals(PJLinkDeviceBindingConstants.CHANNEL_INPUT)
+                    || channelUID.getId().equals(PJLinkDeviceBindingConstants.CHANNEL_INPUT_DYNAMIC)) {
```
You don't need CHANNEL_INPUT_DYNAMIC

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,313 @@
+
+                    StringType input = new StringType(device.getInputStatus().getResult().getValue());
+                    updateState(PJLinkDeviceBindingConstants.CHANNEL_INPUT, input);
+                    updateState(PJLinkDeviceBindingConstants.CHANNEL_INPUT_DYNAMIC, input);
```
You don't need CHANNEL_INPUT_DYNAMIC

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,313 @@
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(PJLinkDeviceConfiguration.class);
```
You should first checks the config, like a non empty IP address or a positive TCP port for example. In case a parameter is not valid, you should set the thing status to OFFLINE with ThingStatusDetail.CONFIGURATION_ERROR.

nils replied:
Do you know if I can rely on e.g. the min/max attributes defined on the thing parameter? 
<parameter name="tcpPort" type="integer" min="0"  max="65535">

lolodomo replied:
No, in case the thing is set in a config file.
But of course the min/max attributes will do the stuff if the user is defining its thing with Paper UI.

nils replied:
Thankfully it's possible to reuse the validation procedure from Paper UI, that's what I'm doing now.

lolodomo commented:
And finally, pjlinkdevice is not so bad ;)

lolodomo commented:
Please don't rename the binding in a first time and push your changes without a force push so that I can easily see what your changes without going again through all your code.

nils commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/device/PJLinkDevice.java:
```diff
@@ -0,0 +1,314 @@
+                if (numberOfTries < 2) {
+                    this.connect(true);
+                } else {
+                    throw e;
```
java.net.SocketException is a subclass of java.io.IOException (which is declared), this should be enough right?

lolodomo replied:
I think you're right.

nils commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/device/PJLinkDevice.java:
```diff
@@ -0,0 +1,314 @@
+                    throw new ResponseException("Invalid header: " + header);
+            }
+        } catch (ConnectException | SocketTimeoutException | NoRouteToHostException e) {
+            throw e;
```
those are subclasses of java.io.IOException (which is declared), this should be enough right?

lolodomo replied:
I think you're right.
But why catching them to then only throw them ?

nils replied:
The first catch is more specific (second catch has IOException), as these Exceptions indicate that there is just no device to connect to (during discovery). I wanted to avoid flooding the log with this information during scan. Except for these three subclasses, IOExceptions seemed to be interesting enough to write a log entry (second catch)

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandlerFactory.java:
```diff
@@ -0,0 +1,101 @@
+ */
+@NonNullByDefault
+@Component(configurationPid = "binding.pjlinkdevice", service = { ThingHandlerFactory.class,
+        ChannelTypeProvider.class }, immediate = true)
```
Note that it was confirmed that "immediate = true" must not be used for a ThingHandlerFactory

nils commented on bundles/org.openhab.binding.pjlinkdevice/.classpath:
```diff
@@ -0,0 +1,50 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
done

nils commented on bundles/org.openhab.binding.pjlinkdevice/.project:
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
done

nils commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,313 @@
+            updateProperty(PJLinkDeviceBindingConstants.PARAMETER_NAME, device.getName());
+        } catch (ResponseException e) {
+            // okay, cannot retrieve model information
+            updateProperty(PJLinkDeviceBindingConstants.PARAMETER_NAME, e.toString());
```
I'd like to get information about whether a device adheres to the PJLink standard or not (and how it deviates). I don't think it's common for users to look at the properties, but during support this information gives me insights how compliant a device is.

lolodomo replied:
Why not a simple log in this case?

lolodomo replied:
Keep as it is if you think it could be really useful for the end user

nils replied:
I'm not sure why I was against just logging initially, as this information intended just for myself (during support requests) and not for the end user, it totally makes sense to just log it.

I'll adjust the code accordingly.

nils commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/device/command/input/Input.java:
```diff
@@ -0,0 +1,120 @@
+        this.getInputType();
+        if (!Arrays.asList(new String[] { "0", "1", "2", "3", "4", "5", "6", "7", "8", "9" })
+                .contains(this.getInputNumber())) {
+            throw new ResponseException("Illegal channel number: " + this.value.substring(1, 2));
```
Could you elaborate a bit?

lolodomo replied:
I mean use "this.getInputNumber()" rather than "this.value.substring(1, 2)"

nils commented on bundles/org.openhab.binding.pjlinkdevice/README.md:
```diff
@@ -0,0 +1,58 @@
+
+## Supported Things
+
+Supported are devices which support the PJLink protocol (class 1). 
```
Me neither, maybe it's better now?

nils commented on bundles/org.openhab.binding.pjlinkdevice/README.md:
````diff
@@ -0,0 +1,58 @@
+
+demo.things:
+
+```java
````
I was under the impression that 'java' is a syntax highlighter that works okay also for OpenHAB syntax. As I don't remember where I got that impression from, I'll remove it.

nils commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/device/command/mute/MuteQueryResponse.java:
```diff
@@ -0,0 +1,84 @@
+        }
+    }
+
+    private MuteQueryResponseValue result = null;
```
This is gone due to the (tiny) refactoring triggered by @NonNullByDefault

lolodomo commented:
Thank for all your changes. I plan to start a new (final) review during the weekend.
What is still missing is the minimum javadoc and the full example in the documentation.

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,323 @@
+    }
+
+    private void clearRefreshInterval() {
+        ScheduledFuture<?> refreshJob = this.refreshJob;
```
Just a remark, not a request for change.
I see you are using the same pattern several times in your code. You do not need to use a local variable. You could directly use the object member.

nils replied:
This pattern had to be introduced when @NonNullByDefault was added.

First, the main reason why @NonNullByDefault makes sense are compiler warnings, so these have to be a clean to gain any benefit. Otherwise (if you just get accustomed to them), you won't notice if there are new ones you need to check.

So without the pattern you noticed, the compiler warns (correctly!) that the variable this.myvar could be null, even if I just checked that it's not null in the line before (due to potential concurrent modifications). The local variable ensures that it cannot have been nulled in the meantime -> no warning -> no hiding of potentially important warnings.

To summarize: @NonNullByDefault does not make sense if you do not also refactor your code to eliminate compiler warnings.

lolodomo replied:
Ok, you're right, I did not know the reason why the compiler was saying the var can be null.

lolodomo commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,323 @@
+
+    @Override
+    public void initialize() {
+        try {
```
Here is how I will do it to avoid a thing remaining in INITIALIZING state in case the setup takes too much time (probably only if the projector is not responding):
```
try {
    this.getDevice();
} catch (ConfigurationException e) {
    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
    return;
}
setupJob = scheduler.schedule() -> {
    setupDevice();
    setupRefreshInterval();  
}, 0, TimeUnit.SECONDS);
```
And of course you cancel setupJob in the dispose() method.

lolodomo replied:
Note that with my code, setupDevice() will never throw ConfigurationException because getDevice() will be called just before.

nils replied:
There are other exception types that can happen during setupDevice() and setupRefreshInterval(), I decided to pull them up and handle all of them at the same level.

nils commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/device/command/ErrorCode.java:
```diff
@@ -0,0 +1,70 @@
+     */
+    public static void checkForErrorStatus(String code, @Nullable Set<ErrorCode> restrictCodesTo) throws ResponseException {
+        ErrorCode parsed = getValueForCode(code);
+        if (parsed != null && (restrictCodesTo == null || restrictCodesTo.contains(parsed))) {
```
That's exactly what I expect. There are error codes ERR1-ERR4. For some commands, e.g. only ERR1-ERR3 are expected error codes, ERR4 would be a valid value (imagine a projector with name "ERR4", this would not be an error but the actual name.

nils replied:
Null means that all error codes are to be expected, so it throws "more" exceptions then if I reduce the potential error codes to be expected.

lolodomo replied:
If you are sure, ok, my feeling was then wrong ;)

nils commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandlerFactory.java:
```diff
@@ -0,0 +1,72 @@
+    @Override
+    protected @Nullable ThingHandler createHandler(Thing thing) {
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+        InputChannelStateDescriptionProvider stateDescriptionProvider = this.stateDescriptionProvider;
```
I will get compiler warnings if I don't, see comment about the pattern. Even though I as a developer know (or maybe only assume?) that there's no concurrent modification possible, the compiler does not know that.

But maybe I can find a more elegant pattern to convince the compiler in this particular case, I'll check that.

nils replied:
Setting the reference already in the constructor makes this part fo the code obsolete.

lolodomo replied:
I hope this is good like you did because this is not the usual way we declare references.

nils commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandlerFactory.java:
```diff
@@ -0,0 +1,72 @@
+@Component(configurationPid = "binding.pjlinkdevice", service = { ThingHandlerFactory.class })
+public class PJLinkDeviceHandlerFactory extends BaseThingHandlerFactory {
+
+    @Nullable
```
Why not @Nullable? I see this as a more readable variant of @NonNullByDefault({}), and there is no need for a 'default' when we're just looking at a single field.

nils replied:
Setting the reference already in the constructor makes this part fo the code obsolete.

nils commented:
Many thanks @lolodomo for your thorough review! I learned quite a few things along the road :-)

nils commented:
Should I squash the changes to one commit now? 69 commits are probably a bit too many just for this little binding :-)

lolodomo commented:
Normally, the maintainer will squash everything when he will merge, no need that you do it yourself.

nils commented:
@Hilbrand: Could you please update your review or approve the changes? I guess there won't be much incentive to merge the binding if there are "Changes requested".

lolodomo commented:
Are there no other controls that could be added?  This is limited.

nils commented:
In which sense 'limited'? The binding implements the PJlink class 1 protocol [as specified](https://pjlink.jbmia.or.jp/english/data_cl2/PJLink_5-1.pdf).

lolodomo commented:
Limited in the sense you can only power on/off and select the source. But ok, other controls, like brightness, conbtrast are probably in another class ?

nils commented:
Class 2 adds a few additional features, see chapters 4.14 to 4.26, as well as a new notification and discovery protocol.

The examples you mentioned are not supported by PJlink. I guess it was already hard enough to standardize a basic set of functionalities across manufacturers.

mstroeve commented:
@Hilbrand any chance you could  update your review or approve the changes?  Because the PR currently is based on outdated "Changes requested" review :-) .

Hilbrand commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,335 @@
+    protected PJLinkDeviceConfiguration getConfiguration() throws ConfigurationException {
+      PJLinkDeviceConfiguration config = this.config;
+      try {
+          this.validateConfigurationParameters(this.getThing().getConfiguration().getProperties());
```
I'm not sure if I understand what your trying to do here. The framework it self already validates the configuration parameters the user set values. So unless you set some configuration values (that are also defined in the thing xml, because that is what is validated against) it make sense to validate here. Otherwise I don't think this is needed.

nils replied:
I noticed that parameters of things defined in .things files did not get validated, that's why I'm calling it manually here. This way the user will get information about parameter validation issues as part of the thing state details.

But I was also confused myself that this was not done by the framework, as it seems like a very common requirement. Maybe I'm doing something wrong, do you have an idea why parameters coming from .things files are not validated?

lolodomo replied:
Maybe they're validated but I think that a binding can be started with a parameter unset while this parameter is declared as required. Generally I add a control in initialize method. Maybe that is useless !  That has to be checked.

lolodomo replied:
And even if the validation is run by the framework, does it lock the binding intialization ? To be checked too.

Hilbrand commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,335 @@
+                    updateState(PJLinkDeviceBindingConstants.CHANNEL_INPUT, input);
+                } else {
+                    logger.trace("Received input command" + command);
+                    Input input = new Input(((StringType) command).toString());
```
You need to check this command is a `StringType`.

nils commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/device/PJLinkDevice.java:
```diff
@@ -0,0 +1,353 @@
+
+                break;
+            } catch (java.net.SocketException e) {
+                if (numberOfTries < 2) {
```
I've refactored that part of the code a bit, the while(true) looked dubious anyways. It's a for loop now, and the dupliction of this.connect() is gone.

nils commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/device/PJLinkDevice.java:
```diff
@@ -0,0 +1,353 @@
+            numberOfTries++;
+            try {
+                socket.getOutputStream().write((fullCommand).getBytes());
+                socket.getOutputStream().flush();
```
I don't think so, the socket is still good for the next request.

nils commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/InputChannelStateDescriptionProvider.java:
```diff
@@ -0,0 +1,62 @@
+ */
+@Component(service = { DynamicStateDescriptionProvider.class, InputChannelStateDescriptionProvider.class })
+@NonNullByDefault
+public class InputChannelStateDescriptionProvider implements DynamicStateDescriptionProvider {
```
Thanks! It's always good to get rid of code :+1:

nils commented:
Thanks @Hilbrand for the updated review!

I've incorporated most of your remarks.

Concerning the reduction of the usage of 'this', I've also removed it for method calls.
You might have guessed that I'm currently mostly working with JavaScript, because you cannot omit 'this' there. 

But in addition to being used to having it everywhere from JS, I even see some advantages of doing it that way also in Java: You always see whether you are dealing with a local variable or with an object field. In that regard, it's similar to the UPPERCASE_CONSTANT naming convention which also shows what you are dealing with without checking the origin explicitly. Especially when working extensively with the nullable annotations, it becomes rather common to have a variable both locally and on the object with the same name, so that is of some value for me.

So if it's okay for you, I'd keep my little thises for instance fields.

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/README.md:
```diff
@@ -0,0 +1,75 @@
+Limitations at this point:
+
+- only IPv4 connections are supported
+- only PJLink Class 1 commands are supported. Class 2 devices should work fine nevertheless, it's just the Class 2 features that will not work.
```
Please use a new line for each sentence

nils replied:
I did so, but this triggers a code analysis warning:
```
[WARNING] .binding.pjlinkdevice/README.md:[15]
The line after a Markdown list must be empty.
```
I guess it's not possible to satisfy both constraints...

J-N-K replied:
Ok. Then better leave it like it is.

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/InputChannelStateDescriptionProvider.java:
```diff
@@ -0,0 +1,40 @@
+      final ChannelTypeI18nLocalizationService channelTypeI18nLocalizationService) {
+    this.channelTypeI18nLocalizationService = null;
+  }
+}
```
Can you check the static code-analysis report (target/code-analysis/report.html) for formating issues? If no report shows up, try rebasing on current master. Older versions of SAT do not show the full report.

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,329 @@
+      PJLinkDevice device = getDevice();
+      switch (channelUID.getId()) {
+        case CHANNEL_POWER:
+          logger.trace("Received power command" + command);
```
```suggestion
          logger.trace("Received power command {}", command);
```
Don't use string concatenation in loggers but placeholders. Also applies to other logging statements.

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,329 @@
+
+      updateStatus(ThingStatus.ONLINE);
+      logger.trace("Successfully handled command {} on channel {}", command, channelUID.getId());
+    } catch (IOException e) {
```
```suggestion
    } catch (IOException | ResponseException e) {
```

and remove the other catch clauses. simpler code is better

nils replied:
Which other catch clauses should I remove? I need them to be able to issue different ThingStatusDetails.

J-N-K replied:
The `ResponseException` immediately below

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,329 @@
+
+  @Override
+  public void initialize() {
+    this.setupJob = scheduler.schedule(() -> {
```
```suggestion
    this.setupJob = scheduler.execute(() -> {
```

nils replied:
The issue with execute() is that it does not return a ScheduledFuture, which means I cannot stop the created job in dispose()/clearSetupJob()

J-N-K replied:
Why would you want to cancel a one-time job?

J-N-K replied:
@nils Can you comment on this one? There are only two points open (the other being the configuration in initialize instead of later method calls) before we can merge. We might get that into M3 then.

lolodomo replied:
In case dispose() is called immediately after initialize().

J-N-K replied:
I doubt this will work. `schedule` with a zero delay will start the task immediately. Calling `cancel` prevents future executions, but does not guarantee that a running task stops. We use `execute` in other places, too. But I would not make my approval dependent on this change.

nils replied:
The need for being able to stop the scheduled connection attempt just started to make sense due to the new autoReconnectPeriod the delay is now also used for delayed reconnections (it's only 0 on first attempt to connect).

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,329 @@
+      } catch (ConfigurationException e) {
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+      }
+    }, 0, TimeUnit.SECONDS);
```
```suggestion
    });
```

If you start immediately anyway, you can use `execute`. 

Question: what if the initial ommunication fails. Will the binding ever recover from that or stay in communication error forever? As simple solution would be to re-schedule the connection attempt after a given time (except in case of configuration error, this will persist until the user changeds the config)

nils replied:
I did not see any binding I use doing that so far. Could you point me to an example that answers the questions arising from that? E.g. which interval to pick, should it be configurable etc...

J-N-K replied:
knx (https://github.com/openhab/openhab2-addons/blob/master/bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/client/AbstractKNXClient.java), onewire (https://github.com/openhab/openhab2-addons/blob/31b9b488adc6b2e2e791e73ba7a11314af1117f3/bundles/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/handler/OwserverBridgeHandler.java#L357)

nils replied:
Thanks, a new autoReconnectPeriod parameter has been introduced.

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,329 @@
+  }
+
+  private void clearSetupJob() {
+    ScheduledFuture<?> setupJob = this.setupJob;
```
```suggestion
    final ScheduledFuture<?> setupJob = this.setupJob;
```

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,329 @@
+  }
+
+  private void clearRefreshInterval() {
+    ScheduledFuture<?> refreshJob = this.refreshJob;
```
```suggestion
    final ScheduledFuture<?> refreshJob = this.refreshJob;
```

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,329 @@
+    try {
+      properties.put(PJLinkDeviceBindingConstants.PROPERTY_NAME, device.getName());
+    } catch (ResponseException e) {
+      logger.debug("Error retrieving property {}, cause: {}", PJLinkDeviceBindingConstants.PROPERTY_NAME, e);
```
```suggestion
      logger.debug("Error retrieving property {}", PJLinkDeviceBindingConstants.PROPERTY_NAME, e);
```

stack trace will be printed anyway if added, also applies to other loggers

nils replied:
You mean it will appear in the log even if I do not provide a placeholder? Could not find any documentation on that...

J-N-K replied:
https://www.baeldung.com/slf4j-log-exceptions

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,329 @@
+    }
+    try {
+      Map<ErrorStatusDevicePart, ErrorStatusQueryResponseState> errorStatus = device.getErrorStatus();
+      for (Map.Entry<ErrorStatusDevicePart, ErrorStatusQueryResponseState> entry : errorStatus.entrySet()) {
```
```suggestion
     errorStatus.forEach((k,v) -> properties.put(PJLinkDeviceBindingConstants.PROPERTY_ERROR_STATUS + k.getCamelCaseText(), v.getText()));
```

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,329 @@
+
+    try {
+      properties.put(PJLinkDeviceBindingConstants.PROPERTY_NAME, device.getName());
+    } catch (ResponseException e) {
```
Is it really necessary to catch all these individually? Or owuld it be possible to add the requested property to the exception when throwing?

nils replied:
I still want to populate the other properties if one fails. So I have to try and catch for each property individually

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandlerFactory.java:
```diff
@@ -0,0 +1,62 @@
+  private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Collections.singleton(THING_TYPE_PJLINK);
+
+  @Activate
+  public PJLinkDeviceHandlerFactory(@Reference InputChannelStateDescriptionProvider provider) {
```
Nice

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/device/PJLinkDevice.java:
```diff
@@ -0,0 +1,348 @@
+        break;
+      } catch (java.net.SocketException e) {
+        closeSocket(socket);
+        socket = null;
```
please remove

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/device/PJLinkDevice.java:
```diff
@@ -0,0 +1,348 @@
+
+  public void dispose() {
+    try {
+      Socket socket = this.socket;
```
```suggestion
      final Socket socket = this.socket;
```

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/device/PJLinkDevice.java:
```diff
@@ -0,0 +1,348 @@
+        closeSocket(socket);
+      }
+    } catch (IOException e) {
+      // okay then, at least we tried
```
Maybe add a trace log or so?

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/device/command/errorstatus/ErrorStatusQueryResponse.java:
```diff
@@ -0,0 +1,118 @@
+  @Override
+  protected Map<ErrorStatusDevicePart, ErrorStatusQueryResponseState> parseResponseWithoutPrefix(
+      String responseWithoutPrefix) throws ResponseException {
+    Map<ErrorStatusDevicePart, ErrorStatusQueryResponseState> result = new HashMap<ErrorStatusDevicePart, ErrorStatusQueryResponseState>();
```
```suggestion
    Map<ErrorStatusDevicePart, ErrorStatusQueryResponseState> result = new HashMap<>();
```

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/device/command/input/Input.java:
```diff
@@ -0,0 +1,128 @@
+    if (obj == null) {
+      return false;
+    }
+    if (getClass() != obj.getClass()) {
```
```suggestion
    if (obj instanceof Input) {
```

since you cast to `Input` later

nils replied:
instanceof would also match subclasses. As this method checks for equality, we can already stop if the two objects are not of the exact same class.

Btw. this is generated code made by the 'generate equals() and hashCode()' wizard of Eclipse, and I think they are typically exactly what is needed :)

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/device/command/input/InputListQueryResponse.java:
```diff
@@ -0,0 +1,50 @@
+ */
+@NonNullByDefault
+public class InputListQueryResponse extends PrefixedResponse<Set<Input>> {
+  private final static HashSet<ErrorCode> SPECIFIED_ERRORCODES = new HashSet<ErrorCode>(
```
```suggestion
  private final static HashSet<ErrorCode> SPECIFIED_ERRORCODES = new HashSet<>(
```

please use diamond operator where possible (also applies to other occurences)

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/discovery/AbstractDiscoveryParticipant.java:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  protected ExecutorService getExecutorService() {
+    ExecutorService executorService = this.executorService;
```
```suggestion
    final ExecutorService executorService = this.executorService;
```

nils replied:
This won't work as I'm setting it to a new instance two lines later.

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/discovery/AbstractDiscoveryParticipant.java:
```diff
@@ -0,0 +1,133 @@
+  protected ExecutorService getExecutorService() {
+    ExecutorService executorService = this.executorService;
+    if (executorService == null) {
+      this.executorService = executorService = Executors
```
Why can't you re-use a shared thread pool here?

Hilbrand replied:
I believe this code pattern used here is used in several other bindings (each copying the code. So maybe it may be a good idea to provide a service in core). It starts several threads scanning the network. I've had discussions about this in the past. The issue is it starts several threads, otherwise it will take far to long to scan the whole network. So the shared scheduler can't be used as it doesn't provide enough concurrent threads. It's probably not the nicest solution, but it seems to do the job.

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,329 @@
+    this.device = null;
+  }
+
+  public void refresh(PJLinkDeviceConfiguration config) {
```
Why do you need config as a parameter here? I would guess that the configuration is not changing at all, so you could just acces the class field.

nils replied:
Directly accessing the class field raises nullability warnings (class field might not have been set yet). 

In case it is not yet set, I'm risking ConfigurationExceptions during creation of one during getConfiguration(). As those would occur in a job, exception handling would be non-trivial.

So in the end, it's easier to give the job what it needs to run as a parameter.

J-N-K replied:
why should that be null if you initialuite the config before in `initialize()`? If you do so, you can add `@NonNullByDefault({})`to the config field.

J-N-K commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java:
```diff
@@ -0,0 +1,329 @@
+
+  private void setupRefreshInterval() throws ConfigurationException {
+    clearRefreshInterval();
+    PJLinkDeviceConfiguration config = PJLinkDeviceHandler.this.getConfiguration();
```
You could move this tto initialize as first call (before scheduling the connect job).

nils replied:
What's the advantage of doing so?

J-N-K replied:
You don't need the config as parameter in the refresh job. if the config changes the handler is disposed and re-initialized anyway, so just accessing fields is faster than parsing/casting config each time you ned a value from that.

nils replied:
I think I understand your point. 

Unfortunately I have to validate the configuration and instantiate the PJLinkDeviceConfiguration at some point. I could do that as early as possible as you propose and rely on having set the field earlier.

However, if I do want to keep nullability warnings to zero, I have the problem that the compiler cannot know whether the field has been set before, even if that seems obvious for me as a developer.

So as an alternative, I have extended getConfiguration() to first check if the validation/instantiation has been done before and reuse the result:
```
protected PJLinkDeviceConfiguration getConfiguration() throws ConfigurationException {
    PJLinkDeviceConfiguration config = this.config;
    if(config != null) {
      return config;
    }
    [... instantiation and validation logic, setting this.config ...]
}
```
This should fix the very valid 'getting and parsing the config over and over again' concern you raised.

However, I still want to pass it to refresh() as a parameter, as I do not want to duplicate the exception handling in case the configuration is invalid. Even if the exception can never be raised at this point (because it was already instantiated earlier during setupDevice()), the method signature of getConfiguration() still contains it and thus it needs to be handled somehow.

Duplicating the exception handling inside of the job can be avoided by calling getConfiguration() already before of the job, so that the exception handling already in place in the stack can be reused. But then I need to pass it to the job as a parameter.

nils commented:
Thanks @J-N-K for the review.

I've fixed most of the issues you noted, but some questions remain (see individual items). I'd be happy for further feedback on these and potentially other issues you notice.

Best Regards,
Nils

nils commented on bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/device/PJLinkDevice.java:
```diff
@@ -0,0 +1,348 @@
+        break;
+      } catch (java.net.SocketException e) {
+        closeSocket(socket);
+        socket = null;
```
Thanks for pointing that out. Looks like keeping 'this' is not the solution to the problem. Do you have a proposal on how to name these local variables acting as a 'threadsave nullable copy'? Maybe there is already an established pattern at OpenHAB for these?

J-N-K commented:
From my POV the config is the only blocking issue that needs to be changed. getting and parsing the config over and over again is not needed, has no benefits and consumes unnecessary resources. It's guaranteed that changing config will re-initialize the handler until the handler defines another behaviour (which is not the case here).

nils commented:
@J-N-K I did my best to follow your advice, however for one review item the solution only partially looks like you proposed. I hope my somewhat lengthy explanation is somehow structured and comprehensible. :smiley:

J-N-K commented:
@nils, I still don't understand what you gain with the null-check. It is guaranteed that `initialize` is called before any other methods (except the constructor),so it is safe to assume `config` is non-null in later usage and you can annotate it with `@NonNullByDefault({})`. If the config validation in `initialize` fails, you can return without setting a new thing status. The framework will then re-try initialization at a later time.

But I'm ok with the state we have now. Thanks for your contribution.

lolodomo commented:
Bravo for the merge.
Bravo to @nils for all his efforts to consider the review comments.
Just one day too late to be included in 2.5 M3 ;)
